[2단계 - 페이먼츠 모듈] 루나(오신정) 미션 제출합니다.

ShinjungOh opened the pull request:
# 🗂️ 페이먼츠 모듈

안녕하세요 온스타! 페이먼츠 모듈 미션 2단계를 완성해서 리뷰 요청 드립니다. 
이번에도 잘 부탁드려요! 🍀

## 🎯 페이먼츠 모듈

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

### 1단계 - Module

- 재사용 가능한 모듈화된 컴포넌트를 개발하고 npm에 배포할 수 있다.
- 재사용 가능한 커스텀 훅을 개발하고 npm에 배포할 수 있다.
- Storybook을 사용하여 컴포넌트의 다양한 상태를 시각적으로 문서화하고 테스트할 수 있다.
- React Testing Library를 사용하여 컴포넌트와 커스텀 훅에 대해 폭넓은 테스트를 작성할 수 있다.

### 2단계 - Refactoring
- 모듈화된 컴포넌트의 재사용성 및 확장성을 경험할 수 있다.
- 요구사항 변경에 따른 컴포넌트 리팩터링 및 개선을 할 수 있다.
- 선택) 실제 프로젝트에서의 컴포넌트 통합 및 활용할 수 있다.


## 🕵️ 셀프 리뷰(Self-Review)

### 제출 전 체크 리스트

- [x] 기능 요구 사항(모달 컴포넌트 및 커스텀 훅 개발, npm 배포)을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] Storybook과 RTL 테스트 케이스를 모두 작성했나요?
- [x] 코드 컨벤션을 지키고, 에러 핸들링을 고려했나요?
- [x] Storybook 데모 페이지에 정상적으로 접근할 수 있나요?
  - [🎨 Storybook 링크](https://6811b32435027dad59e3902b-hdzpptbvyn.chromatic.com/?path=/story/components-modal--center)

### 리뷰 요청 & 논의하고 싶은 내용

### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

`Alert`, `Confirm`, `Prompt` 모달을 구현할 때, Modal.tsx 파일을 활용해 각각의 종류별 모달 파일을 하나씩 만들었습니다(`AlertModal`, `ConfirmModal`, `PromptModal`). 유저가 사용할 때 필요한 모달을 명시적으로 import해서 사용할 수 있도록 하고 싶었어요. (컴포넌트명으로 바로 어떤 모달인지 인지할 수 있게 함.) 
또한 Modal.tsx 파일에서 사이즈와 타입 모두 분기를 나눠 처리하려면 코드가 길어질 것 같아 지금의 방식을 선택했어요. 이렇게 모달 종류별로 파일을 나눈 것이 괜찮은 선택이었는지, 하나의 파일에서 구현하는 것이 나을지, 혹은 제3의 방법이 있는지 궁금합니다.

### 2) 이번 리뷰를 통해 논의하고 싶은 부분

새로운 내용을 학습할 때 어떤 방법으로 하시는지 궁금합니다. 이번 미션을 하면서 `마지막 요소에서 첫 번째 요소로 순환 가능하도록 구현하기` 기능을 `포커스 트랩` 이라고 한다는 것을 알게 되었습니다 이 부분을 어떻게 구현해야 될지 몰라서 AI의 힘을 빌려 `useFocusTrap` 훅을 만들어 사용했는데요, 사실 코드를 다 이해하지는 못했습니다. 

이러한 경우 저는 `1. 그냥 외운다` `2. AI를 활용해 코드를 한 줄 한 줄 분석해 달라고 한다` 이런 방법을 통해 학습해 나갔어요. 그렇지만 머리에 잘 들어오지 않고 어려워서 ‘과연 나 혼자 이 코드를 구현해야 한다면 할 수 있을까’라는 두려움이 생기더라구요. 포커스 트랩을 구현하며 들었던 생각이지만 나아가 이렇게 처음 보는 코드와 구현 방식을 익혀나갈 때 온스타는 어떻게 하셨는지, 어떻게 하고 계시는지 노하우가 궁금합니다.


---

## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도뿐만 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. 모달 컴포넌트의 유연성

- 다양한 위치(중앙, 하단 등)와 내용을 prop으로 받아 모달을 유연하게 구성할 수 있는가?
- 열기, 닫기, 확인 등의 사용자 정의 이벤트 핸들러가 자연스럽게 동작하는가?
- Storybook을 통해 다양한 시나리오를 명확히 테스트했는가?

### 2. 커스텀 훅의 역할과 모듈화

- 카드 정보 입력 유효성 검사 훅이 각각 명확한 책임을 가지도록 분리되어 있는가? 
- 훅이 유효성 검사 결과와 에러 메시지를 명확히 반환하는가?
- 독립적인 테스트 케이스를 작성하여 정상/비정상 입력 모두를 폭넓게 검증했는가?

### 3. 모듈 배포
- npm 배포를 위한 설정이 적절히 구성되어 있는가?

ShinjungOh committed:
docs: 페이먼츠 모듈 README에 요구사항 작성

ShinjungOh committed:
feat: 모달 크기 옵션별 구현(sm, md, lg)

ShinjungOh committed:
feat: AlertModal UI 구현

ShinjungOh committed:
feat: ConfirmModal UI 구현

ShinjungOh committed:
feat: PromptModal UI 구현

ShinjungOh committed:
feat: Button 컴포넌트 분리

ShinjungOh committed:
feat: Alert, Confirm, Prompt 모달 크기 조절

ShinjungOh committed:
feat: 입력 모달 기능 구현

ShinjungOh committed:
feat: Modal CloseButton 자동 포커스 구현

ShinjungOh committed:
chore: Components 디렉토리 생성 및 파일 이동

ShinjungOh committed:
feat: 모달 사이즈에 따른 스토리북 시나리오 작성

ShinjungOh committed:
feat: 스토리북 버튼 타입 오류 해결

ShinjungOh committed:
feat: 모달 컴포넌트 요소 포커스 트랩 구현

ShinjungOh committed:
feat: Alert, Confirm, Prompt 모달 스토리북 추가

ShinjungOh committed:
feat: Alert, Confirm 모달 height 조절

ShinjungOh committed:
docs: 모달 스토리북 사이트 주소 업데이트

ShinjungOh committed:
feat: 카드사별 카드 번호 포맷팅 규칙 적용

ShinjungOh committed:
refactor: 카드 번호 포맷팅 로직 가독성 개선

ShinjungOh committed:
feat: useCardNumber에 코드 포맷터 훅 적용

ShinjungOh committed:
test: useCardFormatter 테스트 코드 작성

ShinjungOh committed:
feat: 카드사 식별 및 유효성 검사 로직 추가

ShinjungOh committed:
refactor: 카드 브랜드 타입 정의 위치 변경

ShinjungOh committed:
refactor: validateCardBrand 유틸 함수 분리 및 useCardBrandValidation에 적용

ShinjungOh committed:
test: useCardBrandValidation 테스트 코드 작성

ShinjungOh committed:
feat: useCardNumber에서 카드 브랜드 유효성 검사 훅 적용

onschan commented on hooks/src/lib/utils/validateCardBrand.ts:
가능하다면 length, reg 등의 필드를 가지는 하나의 객체로 정규식 처리가 조금 더 유리해보이긴하는데, 굳이 반영은 안해주셔도돼요!

ShinjungOh replied:
제가 정규식을 잘 사용하지 못해서, 일부러 미션에서는 최대한 사용을 자제하려고 했어요! 😊

onschan replied:
그랬군요! 정규식도 추후에 한번쯤 학습해보고 LLM도 필요한 잘 만들어주니 잘 활용해보면 좋을 것 같아요!

onschan commented on hooks/src/lib/types/cardTypes.ts:
```diff
@@ -7,3 +7,5 @@ export type CardNumberError = Record<CardNumberKey, string>;
 export type CardExpiryDate = Record<CardPeriodKey, string>;
 export type CardExpiryDateError = Record<CardPeriodKey, string>;
+
+export type CardBrand = 'visa' | 'mastercard' | 'amex' | 'diners' | 'unionpay' | 'unknown';
```
```ts
export const CARD_BRRAND_KEYS = ['visa' ,'mastercard' , 'amex' | 'diners' , 'unionpay' ,'unknown'] as const;

export type CardBrand = typeof CARD_BRRAND_KEYS[number];
```
과 같은 방식으로 관리할 수 있는데,
그보다 'unknown'이라는 타입이 어색해보이는데요.
validate 과정에서 해당 반환값이 존재하기때문에 타입이 선언된걸까요?
없음에 대한 원시타입을 이용해주시는게 어떨까싶어요.

ShinjungOh replied:
말씀하신대로 다시보니 카드 브랜드가 어디에도 속하지 않는 경우를 'unknown'보다는 null로 표현하는 것이 나을 것 같다는 생각이 들었어요. 
명시적으로 '값이 없음'을 표현하려면 null을 사용하기 때문이에요. 
해당 부분과, 커스텀 훅, test 등 관련된 로직을 개선했습니다!

[refactor: 카드 브랜드 식별 불가 상태를 unknown에서 null로 변경](https://github.com/woowacourse/react-modules/pull/153/commits/c1aedba5de25fe59d9cb5efea29d2a8a8ea36a7f)

onschan commented on hooks/src/lib/constants/errorMessages.ts:
as const 처리해주셔도 좋을 것 같아요!

ShinjungOh replied:
`as const`로 선언해주면 `readonly` 프로퍼티를 갖게되고, 해당 문자열 자체가 타입이 되네요. IDE에서도 확인할 수 있어서 편리하네요! 실수로 변경하거나 자동완성이 더 강력해진다는 장점이 있겠어요ㅎㅎ

[refactor: 에러 메시지에 as const 처리](https://github.com/woowacourse/react-modules/pull/153/commits/3bac41f568d58f17ee9882cc12616daa9ca720ad)

### 기존 - as const 사용 전
![CleanShot 2025-05-18 at 20 34 54@2x](https://github.com/user-attachments/assets/f2f70a27-1374-44d6-8678-edf8558e34a9)

### 리팩토링 - as const 사용 후
![CleanShot 2025-05-18 at 20 34 35@2x](https://github.com/user-attachments/assets/4650d752-c432-47ae-ae7a-7bf1eb0aa6a7)

onschan replied:
스크린샷까지 찍어서 보여주셨네요 😄
이런 기회들로 타입시스템의 편리함을 점차 느껴보시면 좋을 것 같아요!

onschan commented on hooks/src/lib/useCardBrandValidation/useCardBrandValidation.ts:
```diff
@@ -0,0 +1,74 @@
+    error: null
+  });
+
+  useEffect(() => {
```
저는 useEffect 내부에서 cardNumber에 의존되어 분기에 따라 상태들을 업데이트해주는건 좋은 패턴은 아니라고 생각해요.

리엑트 공식 문서에서는 [useEffect를 제거할 수 있는 방법](https://ko.react.dev/learn/you-might-not-need-an-effect)을 제공할 정도로 이 사이드이펙트를 통한 상태 업데이트를 지양하도록 하고 있어요.

그 이유는 use-side-effect 라는 이름부터 useEffect는 기존의 리액트 사이클을 벗어나 부수효과를 일으키는 훅이에요. 이 부수효과를 통해 직접 컨트롤할 수 없는 시스템과의 연결, 특정 이벤트 트리거로는 해결하기 복잡한 문제 등을 꽤나 쉽게 해결할 수 있게끔 해주는 훅이기도합니다.

이 부수효과로 이루어진 코드들은 내부에서 선언된 의존성에 대해서만 동작하기때문에 기능의 뎁스가 깊어지고 도메인 레벨이 깊어진다면 디버깅하거나 동작을 유추하기 힘들게 만들 수 있어요.

지금도 useEffect 내부에서도 꽤 많은 분기처리가 이루어지고 있는데, 해당 방식으로 관리가 된다면 유지보수 및 확장에도 불리한 구조로 고착화될 수 있어요.

개선할 수 있는 포인트로 짚어드리면 result 상태 자체는 useState로 선언될 필요가 없는 파생 상태라고 생각해요. 

```
  const [result, setResult] = useState<CardValidationResult>({
    cardBrand: 'unknown',
    isValid: false,
    error: null
  });
```
이 부분을 먼저 지워보시고 다시 한번 로직들을 바라봐주셔도 좋을 것 같아요.

햇갈리실까봐 조금만 더 설명드리면

cardNumber가 변경될떄마다 해당 훅은 재실행되는데요.

useEffect + useState 없이 단순히 변수에 할당해서 반환값으로 관리해도 아무런 문제가 없을거에요.

많은 케이스에서 이런 파생 상태는 직접 상태로 선언하고 관리하는걸 지양하기도하구요. 필연적으로 useEffect가 필요하거든요.

조금 더 나아가보면 이런 변수 할당에 대한 최적화로 useMemo등을 활용해볼 수도 있어요.

ShinjungOh replied:
`파생 상태`라는 용어를 처음 들어서 찾아봤어요. 제가 이런 파생 상태를 여러 개 만들어서 사용하고 있다는 것도 알게 되었어요.
React에서는 최소한의 상태만 만들어야 하고, 가능한 값은 일반 변수나 계산식으로 처리하는 것이 권장되며,
진짜로 `변경을 감지해서 UI를 업데이트해야 하는 값`만 `useState`로 두는 게 좋다고 하네요.

이런 '리액트가 지향하는 바'를 깊이 생각하지 않고 구현할 때 편하고 익숙한 방식으로 상태나 useEffect를 남발했던 것 같아요. 
공식문서도 이 기회에 필요한 항목을 정독해야겠네요!
좋은 인사이트 주셔서 감사합니다🙌

### 파생 상태 

- 파생 상태: props/state 로부터 유추할 수 있는 값을 별도 state 로 저장하는 것
- 단점: 동기화 문제, 불필요한 복잡도, 디버깅 어려움
- 대안:  
	1. 렌더링 시 직접 계산 
	2. 비용이 클 때만 useMemo 사용
- 효과: 코드가 더 단순하고 안정적으로 변화


[refactor: useCardBrandValidation 훅에서 useEffect, useState 제거](https://github.com/woowacourse/react-modules/pull/153/commits/4d5e43c9d23e2de18b489b551ff4d609ade9ddbc)

onschan replied:
잘 학습해주셨고 정리해주신 내용도 좋습니다 😊
최근 Korean FE Article에도 비슷한 내용의 번역글이 올라오기도 했네요. 
[[Korean FE Article] 상태 동기화 함정을 피하는 방법](https://medium.com/@jiwoochoics/%EB%B2%88%EC%97%AD-%EC%83%81%ED%83%9C-%EB%8F%99%EA%B8%B0%ED%99%94-%ED%95%A8%EC%A0%95%EC%9D%84-%ED%94%BC%ED%95%98%EB%8A%94-%EB%B0%A9%EB%B2%95-42096e82d15f)

onschan commented on hooks/src/lib/useCardFormatter/useCardFormatter.ts:
```diff
@@ -0,0 +1,41 @@
+    }, []);
+
+    setFormattedNumber(segments.join(' '));
+  }, [rawNumber, cardBrand]);
```
이 부분도 마찬가지로 useCardBrandValidation 훅에 있는 리뷰와 비슷해요.
rawNumber, formattedNumber 는 파생 상태입니다.

rawNumber를 number 상태로 직접 바라봐주시고, formattedNumber의 상태 선언은 제거해봐주세요.

ShinjungOh replied:
useCardFormatter 훅에서도 위에서 받았던 피드백과 마찬가지로 `useEffect, useState`를 제거했어요. 
캐싱된 계산값이라는 것을 명확히 보여주기 위해 `useMemo`를 사용했는데,
`useMemo`를 사용하기에는 계산 비용이 크지 않은 연산같아서 굳이 사용했나 싶은 생각도 들긴 해요😅

이번 미션을 통해 useState와 useEffect의 불필요한 사용을 줄이고, 정말 필요한 상태는 무엇인지, 어떻게 업데이트 할 수 있을지 최적화하는 연습을 하고 있습니다..!!

해당 코드의 변경으로 관련된 테스트와 type등을 수정해 반영했습니다~!

[refactor: useCardFormatter 훅에서 useEffect, useState 제거](https://github.com/woowacourse/react-modules/pull/153/commits/b53ce6310fdde5b8b916f393cc139bfa8b0bbc8d)
[test: 리팩토링된 커스텀훅에 따른 테스트 코드 수정](https://github.com/woowacourse/react-modules/pull/153/commits/363731fe241985c0e2a40e51e42628d7d945c7f9)

onschan replied:
말씀드린대로 잘 반영해주셨네요!
추가로 useMemo도 캐싱을 위한 메모리 점유가 필요하다보니 계산을 캐싱한다는것이 무조건적으로 좋은 부분은 아니라는건 인식해주시면 좋을 것 같아요 😊

onschan commented on components/src/lib/components/Button/Button.tsx:
```diff
@@ -0,0 +1,29 @@
+  variant: 'confirm' | 'cancel';
+  onClick: () => void;
+  children: React.ReactNode;
+};
```
ButtonHTMLAttributes<HTMLButtonElement> 타입과 rest props 활용해주시면 좋을 것 같아요.

ShinjungOh replied:
`ButtonHTMLAttributes<HTMLButtonElement>` 타입을 상속하고, `rest props`를 활용해 재사용성을 높이고 타입 안전성을 높였어요! 

[refactor: ButtonHTMLAttributes 타입 및 rest props 적용](https://github.com/woowacourse/react-modules/pull/153/commits/8c81031adad877af2f81ad1effd59035a41070e8)

onschan commented on components/src/lib/Modal.tsx:
```diff
@@ -15,8 +18,26 @@ type ModalProps = ModalContainerProps & {
+  const {containerRef, handleKeyDown} = useFocusTrap({
+    initialFocusRef: closeButtonRef as React.RefObject<HTMLButtonElement>,
+    onEscape: onClose
+  });
```
document 레벨이 아니라면 tabIndex가 필요할 수 있을거에요. 
아래 useEffect 제거해보시고 확인해주세요!

ShinjungOh replied:
useEffect를 지우고 useFocusTrap 훅을 이용하는 방법으로 개선했어요!
모달 컨테이너에 `tabIndex={-1}`을 주어, 
모달 컨테이너 자체는 탭 순서에 포함하지 않으면서도
useFocusTrap 훅을 통해 프로그래밍 방식으로 모달 내부의 요소에 포커스를 관리할 수 있도록 구현했어요.

[refactor: Modal 컴포넌트에 tabIndex 속성 추가 및 포커스 관리 개선](https://github.com/woowacourse/react-modules/pull/153/commits/b28b93c72bdad4dd840121f8309b9e9cff6578b1)

onschan replied:
👍👍👍

onschan reviewed (changes requested):
안녕하세요. 루나!
미션 진행하시느라 수고많으셨어요 😊

포커스 트랩 관련해서 실제 동작과 부합하지 않을 수 있는 포인트들이 있긴했는데 말씀해주신 그런 사유들이 있었군요. 😅 리뷰와 답변 확인해주세요!

### 질문에 대한 답변

>새로운 내용을 학습할 때 어떤 방법으로 하시는지 궁금합니다. 이번 미션을 하면서 마지막 요소에서 첫 번째 요소로 순환 가능하도록 구현하기 기능을 포커스 트랩 이라고 한다는 것을 알게 되었습니다 이 부분을 어떻게 구현해야 될지 몰라서 AI의 힘을 빌려 useFocusTrap 훅을 만들어 사용했는데요, 사실 코드를 다 이해하지는 못했습니다.
이러한 경우 저는 1. 그냥 외운다 2. AI를 활용해 코드를 한 줄 한 줄 분석해 달라고 한다 이런 방법을 통해 학습해 나갔어요. 그렇지만 머리에 잘 들어오지 않고 어려워서 ‘과연 나 혼자 이 코드를 구현해야 한다면 할 수 있을까’라는 두려움이 생기더라구요. 포커스 트랩을 구현하며 들었던 생각이지만 나아가 이렇게 처음 보는 코드와 구현 방식을 익혀나갈 때 온스타는 어떻게 하셨는지, 어떻게 하고 계시는지 노하우가 궁금합니다.

지금은 AI가 코딩을 많이 도와주는 세상이라 기능 구현 과정이 더 쉽고 간단할거에요. 이해하지 않아도 만들 수 있는 세상이기도하구요.

사실 AI가 지금처럼 발전하기 이전에도 프로그래밍 세상은 비슷하게 흘러갔어요. 

"프로그래머는 컨트롤 C + 컨트롤 V 만 잘해도된다"는게 이전에도 쭉 있었던 상황이었어요. 관련 개발자 밈도 되게 많았구요. 아이디어가 잘 떠오르지 않는 개발이 필요한 경우 구글링, 스택오버플로우 등등 검색을 통해 개발자 생태계에 남아있는 코드들을 그대로 가져와서 복붙하고 내 환경에 맞게 수정해보고 그런 과정들로 완벽하게 이해하지 못한 기능들을 만들어내고 그랬거든요. 그래서 개발자의 공유 능력과 검색 능력 등등도 굉장히 중요한 요소였구요! AI가 그 코스트를 말도 안되게 낮췄을 뿐이죠.

이 이야기를 먼저 해드린 이유는, AI 사용 시 그 코드를 100% 이해해야된다는 것에서 스트레스는 너무 느끼시지 않았으면 좋다는 의도였어요.
다만 더 발전하고 좋은 개발자가 되기 위해서는 루나가 해주시는 것 처럼 이해하려고 노력해야되는건 물론이죠!

그러면 질문으로 돌아가서 AI 코드 혹은 남이 이미 구현해놓은 코드를 사용할 때 어떻게 이해하는게 좋을까요?
AI가 없던 시절 저같은 경우는 다른 사람의 코드를 이해하는 과정에서 직접 내 코드로 만들어보는 것을 많이 시도했어요.
제 기준에 맞게 변수명도 수정해보고 코드를 조금씩 바꿔보기도 하고 필요없 어보이는 부분은 과감하게 지워버리기도 했구요.
그러다가 동작이 망가지면 아 ‘이 부분은 내가 필요없다고 생각했지만 필요하구나, 왜 달라졌을까? 이건 내가 모르는 영역인게 확실하구나’ 뭐 이런 생각이 자연스럽게 들고 관련된 내용을 검색, 학습하구요.

이 과정을 거치다보면 스스로 기준에 맞는 코드가 만들어질거에요.
직접 수정한 코드를 다시 보면 전체적 흐름을 이해하는데 도움이 됐어요. 그래도 안되면 이건 그냥 이런 경우에 가져와서 쓰자는 식으로 스니펫으로 코드를 저장해놓고 저만의 코드 보물창고를 만들어놓기도했구요.
이런 경험들이 많이 도움이 됐던건 사실인거같아요.

뭐 예전에는 이랬어요. 몇년 되지도 않은 이야기인데 라떼는 이야기처럼 보이기도하네요.

사실 지금은 AI를 좀 더 잘 활용하는 방법을 생각해보는게 더 합리적일거라고 생각하기는 합니다.

말씀하신 `1. 그냥 외운다` 이건 제 기준에서 없는 개념이에요.

코딩은 그냥 외워서 할 수가 없어요. AI가 만들어낸 코드가 정답이 아니에요. 구루 개발자가 짜놓은 코드도 정답이 아니구요. 상황, 환경 모든 케이스마다 적절한 코드와 구현 방식은 매번 달라져요.

`2. AI를 활용해 코드를 한 줄 한 줄 분석해 달라고 한다` 이건 좋은 접근이라고 생각해요.

한 줄 한 줄 분석해가는 것도 좋지만 다른 과정들이 추가가 되면 좋을 것 같아요.

먼저, AI에게 `왜` 이렇게 했는지 물어보는거에요.

포커스 트랩 기능은 왜 필요한지,  구현을 위해 왜 이런 코드를 만들어냈는지, 기능 구현을 위해 어떤게 필요하길래, `'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])'` 이런 요소들을 왜 찾아서 할당하는지, 왜 keydown 이벤트를 쓴건지, 왜 escape와 tap 등을 가려내는지 등등요. 이해가 되지 않은 상태라면 궁금한 점이 꽤 많을거에요. AI는 이런 질문에 대해서 답변을 잘하는 편이구요.

이런 괒어으로 궁금한 부분들이 어느정도 풀렸다면, 다음엔 `다른 구현 방법은 어떤게 있는지` 를 물어보는거에요. 될 수 있다면 많이요.

그리고 이전 방법과 새로운 방법은 어떤 차이점이 있는지, 어떤 케이스에 선택해서 쓰면 좋은지 이런 부분들도 물어보면 좋아요. 그러면 완벽히 코드를 이해하지 않더라도 이런 케이스엔 이런 식으로 코드 구성을 할 수는 있겠구나 하는 감은 분명히 올거에요.

저한테는 이런 루틴이 도움이 많이 되더라구요. 

굉장히 현실적인 고민을 공유해주셔서 저도 주절주절 적어버렸는데, 도움이 되었으면 좋겠네요. 궁금한 점 있다면 다시 질문해주세요!

ShinjungOh commented:
@onschan 너무나 따뜻하고 정성스러운 답변 감사합니다...🥹 개발 공부를 하다가 어느 순간 AI가 대유행하기 시작했고, 그 과정에서 제가 학습에 AI를 사용하는 것에 은연중에 죄책감을 느껴왔던 것 같아요. 구글링을 하면서 다른 사람의 코드를 찾아보는건 괜찮은데, 왜 AI는 사용하기 껄끄러웠을까 생각해보면 구글링할 때 드는 에너지 자체를 '학습하고 있다'고 느꼈던 것 같기도 하네요. 말씀하신대로 `AI가 그 코스트가 말도 안되게 낮췄을 뿐`인데 말이에요ㅎㅎ

갑자기 저의 고민 상담 시간이 되었는데 같이 고민해주시고 경험 공유해주셔서 감사해요.
알려주신 학습 방법 참고해서 `스스로 기준에 맞는 코드`를 만들어가도록 할게요!! 🥰

지금 페어 프로그래밍 기간이라 페어 미션이 급한 관계로, 페이먼츠 모듈 미션은 이번주 일요일까지 리팩토링 마치고 요청 보내겠습니다! 
좋은 하루 되세요~!

ShinjungOh committed:
refactor: 에러 메시지에 as const 처리

ShinjungOh committed:
refactor: ButtonHTMLAttributes 타입 및 rest props 적용

ShinjungOh committed:
refactor: 모달 컴포넌트 스토리 코드의 상태 변수명 통일

ShinjungOh committed:
refactor: 모달 스토리의 중복 타입 제거 및 사이즈별 크기 스타일 코드 재활용

ShinjungOh commented on components/src/lib/components/Modal/AlertModal.stories.tsx:
```diff
@@ -0,0 +1,56 @@
+        )}
+      </>
+    );
+  },
```
[refactor: 모달 컴포넌트 스토리 코드의 상태 변수명 통일](https://github.com/woowacourse/react-modules/pull/153/commits/53469cf39d85a945b7dcbb48cbcd025d26425326)

ShinjungOh commented on components/src/lib/Modal.stories.tsx:
```diff
@@ -1,8 +1,26 @@
+    width: '600px',
+    height: '206px',
+  },
+};
```
중복된 타입을 삭제하고, 모달 사이즈별 크기를 `Modal.styles.ts` 파일에서 가져다 사용하도록 분리했어요~

[refactor: 모달 스토리의 중복 타입 제거 및 사이즈별 크기 스타일 코드 재활용](https://github.com/woowacourse/react-modules/pull/153/commits/069495a711231e9804d258714a50ef2de4277ad4)

ShinjungOh committed:
refactor: 스토리북 메타데이터 계층 구조 개선

ShinjungOh commented on components/src/lib/components/Modal/ConfirmModal.stories.tsx:
```diff
@@ -0,0 +1,53 @@
+      },
+    },
+  },
+};
```
[refactor: 스토리북 메타데이터 계층 구조 개선](https://github.com/woowacourse/react-modules/pull/153/commits/c65f321a49d6c7e5f963ef931a345bc2020c8ccd)

### 스토리북 목록 스크린샷

![CleanShot 2025-05-18 at 23 50 43@2x](https://github.com/user-attachments/assets/e1f44b0c-6f4f-49d3-8206-b5ea075cdc89)

ShinjungOh commented on hooks/src/lib/useCardBrandValidation/useCardBrandValidation.ts:
```diff
@@ -0,0 +1,74 @@
+  });
+
+  useEffect(() => {
+    const digits = cardNumber.split(' ').join('');
+    const isAllDigits = [...digits].every((num) => num >= '0' && num <= '9');
+
+    if (digits === '') {
+      setResult({
+        cardBrand: 'unknown',
+        isValid: false,
+        error: CARD_NUMBER_ERROR.required,
+      });
+      return;
+    }
+
+    if (!isAllDigits) {
+      setResult({
+        cardBrand: 'unknown',
+        isValid: false,
+        error: CARD_NUMBER_ERROR.onlyNumbers,
+      });
+      return;
+    }
```
정규식 사용 방법을 잘 알지 못해서, 일부러 미션에서는 최대한 사용을 자제하려고 했어요!

ShinjungOh committed:
refactor: 카드 브랜드 식별 불가 상태를 unknown에서 null로 변경

ShinjungOh committed:
refactor: useCardBrandValidation 훅에서 useEffect, useState 제거

ShinjungOh committed:
refactor: useCardFormatter 훅에서 useEffect, useState 제거

- 카드 번호 상태 관리 최적화 및 포맷팅 로직 개선

ShinjungOh committed:
test: 리팩토링된 커스텀훅에 따른 테스트 코드 수정

- useCardFormatter, useCardNumber, useCardExpiryPeriod 테스트 코드 리팩토링
- cardTypes에서 CardNumberError, CardExpiryDateError의 타입에 null 추가

ShinjungOh committed:
docs: README의 모달 스토리북 배포 주소 변경

ShinjungOh committed:
refactor: Modal 컴포넌트에 tabIndex 속성 추가 및 포커스 관리 개선

ShinjungOh commented:
# 📦 페이먼츠 모듈 - 리팩토링

안녕하세요 온스타! 주말 잘 보내셨나요?ㅎㅎ 
피드백 반영을 마치고 리뷰 요청 드려요. 
리팩토링하며 변경된 사항과 관련 자료들 같이 정리해 올립니다!

모달을 구현하면서 재활용하기 좋은 유연한 모달이란 무엇인지 고민하는 시간이 되었어요.
커스텀훅을 개선하면서 리액트의 지향점을 다시 한 번 떠올리며, 기존에 갖고있던 저의 코드 습관을 어떻게 개선해야 할지를 알게 되었고,
공식문서를 정독해야겠다는 의지를 불태워주었습니다. 🔥 
페이먼츠 미션은 한 달 동안 진행되기도 했고, 특히 이번 모듈과 같은 경우에는 화면을 그리는 것이 없다보니 지치기도 한 미션이었는데요,
시각을 바꿔서 리액트의 훅을 딥다이브할 수 있는 기회였다는 점에서 막판에 재미가 붙었던 것 같아요ㅎㅎ 
알려주신 공식 문서 자료 꼭 학습해 보도록 하겠습니다!!

[🎨 스토리북 배포 주소](https://68009db56ddb7f7e2eb2c11e-mkauyxvvpe.chromatic.com/?path=/story/components-button--rounded)

<br>

## 📹 모달 테스트 영상 

### 컨펌 모달
https://github.com/user-attachments/assets/1efdbb21-eeed-4624-ab63-e71d8431624a


### 프롬프트 모달
https://github.com/user-attachments/assets/e26983a5-34c0-4a3d-aa17-6fb442374a95

<br>

## 💯 커스텀훅 테스트 결과

<img src='https://github.com/user-attachments/assets/0b33c8be-f2a2-4e86-9552-a3ce85fdaa72' width='550px' alt='테스트결과'/>


<br>

## ⚙️ 주요 변경 사항

### 1. 커스텀훅의 불필요한 useEffect, useState 제거 

이번 리팩토링에서 집중적으로 신경쓴 부분입니다. 
`파생 상태`라는 키워드를 새로 배웠고, 습관적으로 남용하던 `useEffect, useState`를 사용하지 않고도 구현하는 방법을 학습했어요.
사용자의 입력에 해당하지 않는 카드 브랜드가 있다면 'unknown' 대신 null을 사용하도록 했고, 
해당 부분의 리팩토링으로 인해 관련된 테스트 코드를 수정했어요.

[refactor: useCardBrandValidation 훅에서 useEffect, useState 제거](https://github.com/woowacourse/react-modules/pull/153/commits/4d5e43c9d23e2de18b489b551ff4d609ade9ddbc)
[refactor: useCardFormatter 훅에서 useEffect, useState 제거](https://github.com/woowacourse/react-modules/pull/153/commits/b53ce6310fdde5b8b916f393cc139bfa8b0bbc8d)
[test: 리팩토링된 커스텀훅에 따른 테스트 코드 수정](https://github.com/woowacourse/react-modules/pull/153/commits/363731fe241985c0e2a40e51e42628d7d945c7f9)


### 2. 모달 컴포넌트 포커스 관리 및 타입 개선

`useFocusTrap` 훅을 활용해 Modal 컴포넌트에서 useEffect를 사용하지 않고 포커스 관리를 할 수 있도록 개선했어요.
버튼과 에러 메시지 등에서 피드백을 바탕으로 한 타입 개선이 이루어졌어요.

[refactor: Modal 컴포넌트에 tabIndex 속성 추가 및 포커스 관리 개선](https://github.com/woowacourse/react-modules/pull/153/commits/b28b93c72bdad4dd840121f8309b9e9cff6578b1)
[refactor: 에러 메시지에 as const 처리](https://github.com/woowacourse/react-modules/pull/153/commits/3bac41f568d58f17ee9882cc12616daa9ca720ad)
[refactor: ButtonHTMLAttributes 타입 및 rest props 적용](https://github.com/woowacourse/react-modules/pull/153/commits/8c81031adad877af2f81ad1effd59035a41070e8)


### 3. 모달 컴포넌트 스토리 개선

코드래빗의 리뷰를 참고해, 스토리 코드에서 개선할 수 있는 포인트를 반영했어요. 
크로마틱으로 스토리북을 배포해 최신 링크를 리드미에 반영했어요.

[refactor: 모달 컴포넌트 스토리 코드의 상태 변수명 통일](https://github.com/woowacourse/react-modules/pull/153/commits/53469cf39d85a945b7dcbb48cbcd025d26425326)
[refactor: 모달 스토리의 중복 타입 제거 및 사이즈별 크기 스타일 코드 재활용](https://github.com/woowacourse/react-modules/pull/153/commits/069495a711231e9804d258714a50ef2de4277ad4)
[refactor: 스토리북 메타데이터 계층 구조 개선](https://github.com/woowacourse/react-modules/pull/153/commits/c65f321a49d6c7e5f963ef931a345bc2020c8ccd)
[docs: README의 모달 스토리북 배포 주소 변경](https://github.com/woowacourse/react-modules/pull/153/commits/ff949b0c7243c8af4e31543451a714887765fd16)

onschan reviewed (approved):
제가 말씀드린 부분 스펀지처럼 쏙쏙 잘 흡수하시고 학습도 잘해주셨네요! 아주 좋습니다.
일을 하시다보면 단순 화면만 그리고 기능만 붙이는것 이외에도 아키텍처 차원의 고민과 구조적인 고민에 대한 일을 하게되는 순간들이 많이 찾아올거에요!
단순 리액트 프레임워크를 벗어나서 웹 생태계에 대한 전반적인 이해가 필요한 순간들일거구요!
하나의 작업을 진행하시더라도 다양한 시각으로 바라보고 작업해주신다면 추후에 분명 큰 도움이 될거에요.
미션 진행하시느라 수고많으셨어요. 이번 미션은 여기서 Aprrove하도록 하겠습니다!
